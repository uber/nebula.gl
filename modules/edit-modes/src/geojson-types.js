// @flow

// Geometry types

export type Position = [number, number] | [number, number, number];
export type PointCoordinates = Position;
export type LineStringCoordinates = Position[];
export type PolygonCoordinates = Position[][];
export type MultiPointCoordinates = Position[];
export type MultiLineStringCoordinates = Position[][];
export type MultiPolygonCoordinates = Position[][][];

export type AnyCoordinates =
  | PointCoordinates
  | LineStringCoordinates
  | PolygonCoordinates
  | MultiPointCoordinates
  | MultiLineStringCoordinates
  | MultiPolygonCoordinates;

export type Point = {
  type: 'Point',
  coordinates: PointCoordinates
};

export type LineString = {
  type: 'LineString',
  coordinates: LineStringCoordinates
};

export type Polygon = {
  type: 'Polygon',
  coordinates: PolygonCoordinates
};

export type MultiPoint = {
  type: 'MultiPoint',
  coordinates: MultiPointCoordinates
};

export type MultiLineString = {
  type: 'MultiLineString',
  coordinates: MultiLineStringCoordinates
};

export type MultiPolygon = {
  type: 'MultiPolygon',
  coordinates: MultiPolygonCoordinates
};

export type Geometry = Point | LineString | Polygon | MultiPoint | MultiLineString | MultiPolygon;

export type Polygonal = Polygon | MultiPolygon;

// Feature types

export type BoundingBoxArray = [number, number, number, number];

export type FeatureOf<T: Geometry> = {
  type: 'Feature',
  geometry: T,
  properties?: {},
  id?: string | number,
  bbox?: BoundingBoxArray
};

export type FeatureWithProps<T: Geometry, P> = {
  type: 'Feature',
  geometry: T,
  properties: P
};

export type Feature =
  | FeatureOf<Point>
  | FeatureOf<LineString>
  | FeatureOf<Polygon>
  | FeatureOf<MultiPoint>
  | FeatureOf<MultiLineString>
  | FeatureOf<MultiPolygon>;

export type FeatureCollection = {
  type: 'FeatureCollection',
  features: Feature[],
  properties?: {},
  id?: string | number,
  bbox?: BoundingBoxArray
};

export type AnyGeoJson = Feature | FeatureCollection;
