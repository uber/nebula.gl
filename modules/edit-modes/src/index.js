// @flow

export { EditMode, BaseEditMode } from './lib/edit-mode.js';
export { ViewMode } from './lib/view-mode.js';
export { DrawPolygonMode } from './lib/draw-polygon-mode.js';
export { ImmutableFeatureCollection } from './lib/immutable-feature-collection.js';

export type { ModeState } from './lib/edit-mode.js';

export type {
  ScreenCoordinates,
  EditAction,
  Pick,
  ClickEvent,
  PointerMoveEvent,
  StartDraggingEvent,
  StopDraggingEvent
} from './types.js';

export type {
  Position,
  PointCoordinates,
  LineStringCoordinates,
  PolygonCoordinates,
  MultiPointCoordinates,
  MultiLineStringCoordinates,
  MultiPolygonCoordinates,
  AnyCoordinates,
  Point,
  LineString,
  Polygon,
  MultiPoint,
  MultiLineString,
  MultiPolygon,
  Geometry,
  Polygonal,
  BoundingBoxArray,
  FeatureOf,
  FeatureWithProps,
  Feature,
  FeatureCollection,
  AnyGeoJson
} from './geojson-types.js';
