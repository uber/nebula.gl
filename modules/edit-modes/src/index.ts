export type { EditMode } from './lib/edit-mode';
export type { GeoJsonEditModeType } from './lib/geojson-edit-mode';

export { GeoJsonEditMode } from './lib/geojson-edit-mode';

// Alter modes
export { ModifyMode } from './lib/modify-mode';
export { TranslateMode } from './lib/translate-mode';
export { ScaleMode } from './lib/scale-mode';
export { RotateMode } from './lib/rotate-mode';
export { DuplicateMode } from './lib/duplicate-mode';
export { ExtendLineStringMode } from './lib/extend-line-string-mode';
export { SplitPolygonMode } from './lib/split-polygon-mode';
export { ExtrudeMode } from './lib/extrude-mode';
export { ElevationMode } from './lib/elevation-mode';
export { TransformMode } from './lib/transform-mode';

// Draw modes
export { DrawPointMode } from './lib/draw-point-mode';
export { DrawLineStringMode } from './lib/draw-line-string-mode';
export { DrawPolygonMode } from './lib/draw-polygon-mode';
export { DrawRectangleMode } from './lib/draw-rectangle-mode';
export { DrawCircleByDiameterMode } from './lib/draw-circle-by-diameter-mode';
export { DrawCircleFromCenterMode } from './lib/draw-circle-from-center-mode';
export { DrawEllipseByBoundingBoxMode } from './lib/draw-ellipse-by-bounding-box-mode';
export { DrawEllipseUsingThreePointsMode } from './lib/draw-ellipse-using-three-points-mode';
export { DrawRectangleUsingThreePointsMode } from './lib/draw-rectangle-using-three-points-mode';
export { Draw90DegreePolygonMode } from './lib/draw-90degree-polygon-mode';
export { DrawPolygonByDraggingMode } from './lib/draw-polygon-by-dragging-mode';
export { ImmutableFeatureCollection } from './lib/immutable-feature-collection';

// Other modes
export { ViewMode } from './lib/view-mode';
export { MeasureDistanceMode } from './lib/measure-distance-mode';
export { MeasureAreaMode } from './lib/measure-area-mode';
export { MeasureAngleMode } from './lib/measure-angle-mode';
export { CompositeMode } from './lib/composite-mode';
export { SnappableMode } from './lib/snappable-mode';

// Experimental
export {default as _memoize} from './memoize';

export type {
  ScreenCoordinates,
  EditAction,
  Pick,
  ClickEvent,
  PointerMoveEvent,
  StartDraggingEvent,
  StopDraggingEvent,
  DraggingEvent,
  DoubleClickEvent,
  ModeProps,
  GuideFeatureCollection,
  Viewport,
  Tooltip,
} from './types';

export type {
  Position,
  PointCoordinates,
  LineStringCoordinates,
  PolygonCoordinates,
  MultiPointCoordinates,
  MultiLineStringCoordinates,
  MultiPolygonCoordinates,
  AnyCoordinates,
  Point,
  LineString,
  Polygon,
  MultiPoint,
  MultiLineString,
  MultiPolygon,
  Geometry,
  Polygonal,
  BoundingBoxArray,
  FeatureOf,
  FeatureWithProps,
  Feature,
  FeatureCollection,
  AnyGeoJson,
} from './geojson-types';
