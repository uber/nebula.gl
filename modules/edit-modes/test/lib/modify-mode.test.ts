/* eslint-env jest */

import { ModifyMode } from '../../src/lib/modify-mode';
import { Pick, ModeProps } from '../../src/types';
import {
  createFeatureCollectionProps,
  createPointerMoveEvent,
  createStartDraggingEvent,
  createStopDraggingEvent,
} from '../test-utils';
import { FeatureCollection, Position, Point, LineString, FeatureOf } from '../../src/geojson-types';

let pointFeature: FeatureOf<Point>;
let lineStringFeature: FeatureOf<LineString>;
let polygonFeature;
let polygonRectangleFeature;
let multiPointFeature;
let multiLineStringFeature;
let multiPolygonFeature;

beforeEach(() => {
  pointFeature = {
    type: 'Feature',
    properties: {},
    geometry: { type: 'Point', coordinates: [1, 2] },
  };

  lineStringFeature = {
    type: 'Feature',
    properties: {},
    geometry: {
      type: 'LineString',
      coordinates: [
        [1, 2],
        [2, 3],
        [3, 4],
      ],
    },
  };

  polygonFeature = {
    type: 'Feature',
    properties: {},
    geometry: {
      type: 'Polygon',
      coordinates: [
        // exterior ring
        [
          [-1, -1],
          [1, -1],
          [1, 1],
          [-1, 1],
          [-1, -1],
        ],
        // hole
        [
          [-0.5, -0.5],
          [-0.5, 0.5],
          [0.5, 0.5],
          [0.5, -0.5],
          [-0.5, -0.5],
        ],
      ],
    },
  };

  polygonRectangleFeature = {
    type: 'Feature',
    properties: { shape: 'Rectangle' },
    geometry: {
      type: 'Polygon',
      coordinates: [
        [
          [-100, -100],
          [100, -100],
          [100, 100],
          [-100, 100],
          [-100, -100],
        ],
      ],
    },
  };

  multiPointFeature = {
    type: 'Feature',
    properties: {},
    geometry: {
      type: 'MultiPoint',
      coordinates: [
        [1, 2],
        [3, 4],
      ],
    },
  };

  multiLineStringFeature = {
    type: 'Feature',
    properties: {},
    geometry: {
      type: 'MultiLineString',
      coordinates: [
        [
          [1, 2],
          [2, 3],
          [3, 4],
        ],
        [
          [5, 6],
          [6, 7],
          [7, 8],
        ],
      ],
    },
  };

  multiPolygonFeature = {
    type: 'Feature',
    properties: {},
    geometry: {
      type: 'MultiPolygon',
      coordinates: [
        [
          // exterior ring polygon 1
          [
            [-1, -1],
            [1, -1],
            [1, 1],
            [-1, 1],
            [-1, -1],
          ],
          // hole  polygon 1
          [
            [-0.5, -0.5],
            [-0.5, 0.5],
            [0.5, 0.5],
            [0.5, -0.5],
            [-0.5, -0.5],
          ],
        ],
        [
          // exterior ring polygon 2
          [
            [2, -1],
            [4, -1],
            [4, 1],
            [2, 1],
            [2, -1],
          ],
        ],
      ],
    },
  };
});

const mockMove = (mode, picks: Pick[], props: ModeProps<FeatureCollection>) => {
  const moveEvent = createPointerMoveEvent([100, 100], picks);
  mode.handlePointerMove(moveEvent, props);

  const startDragEvent = createStartDraggingEvent([100, 100], [100, 100], picks);
  mode.handleStartDragging(startDragEvent, props);

  const stopDragEvent = createStopDraggingEvent([110, 115], [100, 100], picks);
  mode.handleStopDragging(stopDragEvent, props);
};

test('Rectangular polygon feature preserves shape', () => {
  const mockOnEdit = jest.fn();
  const props = createFeatureCollectionProps({
    data: {
      type: 'FeatureCollection',
      features: [polygonRectangleFeature],
    } as FeatureCollection,
    selectedIndexes: [0],
    onEdit: mockOnEdit,
  });

  const mode = new ModifyMode();
  const guides = mode.getGuides(props);
  expect(guides).toMatchSnapshot();

  const guideFeature = guides.features[2];
  mockMove(mode, [{ index: 2, isGuide: true, object: guideFeature }], {
    ...props,
    modeConfig: { lockRectangles: true },
  });

  expect(mockOnEdit).toHaveBeenCalledTimes(1);
  const movedFeature = mockOnEdit.mock.calls[0][0].updatedData.features[0];
  expect(movedFeature).toMatchSnapshot();
  expect(props.data.features[0]).not.toEqual(movedFeature);
});

describe('getGuides()', () => {
  it('gets edit handles for Point', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [pointFeature],
      } as FeatureCollection,
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  it('gets edit handles for LineString', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [lineStringFeature],
      },
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  it('gets edit handles for Polygon', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [polygonFeature],
      },
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  it('gets edit handles for MultiPoint', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [multiPointFeature],
      },
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  it('gets edit handles for MultiLineString', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [multiLineStringFeature],
      },
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  it('gets edit handles for MultiPolygon', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [multiPolygonFeature],
      },
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  it('gets edit handles for all selected features in collection', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [lineStringFeature, pointFeature, multiPointFeature],
      } as FeatureCollection,
      selectedIndexes: [0, 2],
    });

    const guides = mode.getGuides(props);

    expect(guides).toMatchSnapshot();
  });

  const lineString: FeatureOf<LineString> = {
    type: 'Feature',
    geometry: {
      type: 'LineString',
      coordinates: [
        [-122.40880966186523, 37.783536601521924],
        [-122.43893623352051, 37.779669924659004],
        [-122.43515968322752, 37.7624370109886],
        [-122.42348670959471, 37.77180027337861],
        [-122.4250316619873, 37.778584505321376],
        [-122.42314338684082, 37.778652344496926],
        [-122.42357254028322, 37.77987343901049],
        [-122.41198539733887, 37.78109451335266],
      ],
    },
  };

  const point = {
    type: 'Feature',
    geometry: {
      type: 'Point',
      coordinates: [-122.40880966186523, 37.783536601521924],
    },
  };
  const pick = {
    object: lineString,
    isGuide: false,
    index: 0,
  };

  const mapCoords: Position = [-122.43862233312133, 37.77767798407437];

  it('includes an intermediate edit handle', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [lineString],
      } as FeatureCollection,
      selectedIndexes: [0],
      lastPointerMoveEvent: {
        picks: [pick],
        mapCoords,
        screenCoords: [42, 42],
        cancelPan: jest.fn(),
        sourceEvent: null,
      },
    });

    const guides = mode.getGuides(props);

    const intermediate = guides.features.find(
      ({ properties }) =>
        properties.guideType === 'editHandle' && properties.editHandleType === 'intermediate'
    );

    expect(intermediate).toMatchSnapshot();
  });

  it('does not add intermeidate edit handle when no picks provided', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [lineString],
      },
      selectedIndexes: [0],
    });

    const guides = mode.getGuides(props);

    const intermediate = guides.features.find(
      ({ properties }) =>
        properties.guideType === 'editHandle' && properties.editHandleType === 'intermediate'
    );
    expect(intermediate).toBeUndefined();
  });

  it('does not add intermeidate edit handle when too close to existing edit handle', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [lineString],
      } as FeatureCollection,
      selectedIndexes: [0],
      lastPointerMoveEvent: {
        picks: [
          pick,
          {
            isGuide: true,
            index: 42,
            object: {
              properties: { guideType: 'editHandle', editHandleType: 'existing', featureIndex: 0 },
              geometry: { coordinates: [] },
            },
          },
        ],
        mapCoords,
        screenCoords: [42, 42],
        cancelPan: jest.fn(),
        sourceEvent: null,
      },
    });

    const guides = mode.getGuides(props);

    const intermediate = guides.features.find(
      ({ properties }) =>
        properties.guideType === 'editHandle' && properties.editHandleType === 'intermediate'
    );
    expect(intermediate).toBeUndefined();
  });

  it('does not add intermeidate edit handle when pick is not a selected feature', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [lineString],
      } as FeatureCollection,
    });

    const guides = mode.getGuides(props);

    const intermediate = guides.features.find(
      ({ properties }) =>
        properties.guideType === 'editHandle' && properties.editHandleType === 'intermediate'
    );
    expect(intermediate).toBeUndefined();
  });

  it('does not add intermeidate edit handle when pick is a Point / MultiPoint', () => {
    const mode = new ModifyMode();
    const props = createFeatureCollectionProps({
      data: {
        type: 'FeatureCollection',
        features: [point],
      } as FeatureCollection,
      selectedIndexes: [0],
      lastPointerMoveEvent: {
        picks: [
          {
            isGuide: false,
            index: 0,
            object: point,
          },
        ],
        mapCoords,
        screenCoords: [42, 42],
        cancelPan: jest.fn(),
        sourceEvent: null,
      },
    });
    const guides = mode.getGuides(props);
    const intermediate = guides.features.find(
      ({ properties }) =>
        properties.guideType === 'editHandle' && properties.editHandleType === 'intermediate'
    );
    expect(intermediate).toBeUndefined();
  });
});
