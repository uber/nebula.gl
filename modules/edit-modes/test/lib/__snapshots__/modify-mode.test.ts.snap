// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rectangular polygon feature preserves shape 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          -100,
          -100,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          100,
          -100,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          100,
          100,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -100,
          100,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          3,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`Rectangular polygon feature preserves shape 2`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        Array [
          -100,
          -100,
        ],
        Array [
          -100,
          115,
        ],
        Array [
          110,
          115,
        ],
        Array [
          110,
          -100,
        ],
        Array [
          -100,
          -100,
        ],
      ],
    ],
    "type": "Polygon",
  },
  "properties": Object {
    "shape": "Rectangle",
  },
  "type": "Feature",
}
`;

exports[`getGuides() gets edit handles for LineString 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          2,
          3,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          3,
          4,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          2,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() gets edit handles for MultiLineString 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          2,
          3,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          3,
          4,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          5,
          6,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          6,
          7,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          7,
          8,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          2,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() gets edit handles for MultiPoint 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          3,
          4,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() gets edit handles for MultiPolygon 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          -1,
          -1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          0,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          -1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          0,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          0,
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -1,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          0,
          3,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -0.5,
          -0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          1,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -0.5,
          0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          1,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          0.5,
          0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          1,
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          0.5,
          -0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          1,
          3,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          2,
          -1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          0,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          4,
          -1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          0,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          4,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          0,
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          2,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          0,
          3,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() gets edit handles for Point 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() gets edit handles for Polygon 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          -1,
          -1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          -1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -1,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
          3,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -0.5,
          -0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -0.5,
          0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          0.5,
          0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          0.5,
          -0.5,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
          3,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() gets edit handles for all selected features in collection 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          2,
          3,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          3,
          4,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 0,
        "guideType": "editHandle",
        "positionIndexes": Array [
          2,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 2,
        "guideType": "editHandle",
        "positionIndexes": Array [
          0,
        ],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          3,
          4,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": 2,
        "guideType": "editHandle",
        "positionIndexes": Array [
          1,
        ],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`getGuides() includes an intermediate edit handle 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      -122.4385053241134,
      37.77770362630828,
    ],
    "type": "Point",
  },
  "properties": Object {
    "editHandleType": "intermediate",
    "featureIndex": 0,
    "guideType": "editHandle",
    "positionIndexes": Array [
      2,
    ],
  },
  "type": "Feature",
}
`;
